Add Ensemble Objects

ALescoulie opened the pull request:
# Adding Ensemble and Ensemble Analysis

Created a new pull request just for the ensemble object so that the history would be less chaotic. I will build and commit the analysis classes in another fork and pull request. My previous draft request #169 was branched from my python 3 support branch making the commit history a nightmare, so I just created a new one. 

I also decided to separate the ensemble objects into their own file.

## Summary of New Objects

### Ensemble

The Ensemble object is a collection of [MDAnalyis](https://github.com/MDAnalysis/mdanalysis) Universe objects. It is intended to store the set of systems generated by running mdpow-fep. 

The Ensemble object works by storing the systems in a dictionary and extending the functionality of a Universe object to a collection of universes. It when given a directory finds the simulation files, reads then loads them into a dictionary. The object can be indexed the same as a dictionary, and has methods analogous the the Universe object. The main one being `select_atoms` which returns a *EnsembleAtomGroup*.

An Ensemble in its current form can also be built by manually adding and popping universes into an empty instance.

```python
benzene_dir = os.path.join('mdpow', 'tests', 'testing_resources', 'states', 'benzene')

Benzene = mdpow.ensemble.Ensemble(dirname=benzene_dir)

DihedralGroup = Benzene.select_atoms('name C1 or name C2 or name C3 or name C4')
```

### EnsembleAtomGroup

The EnsembleAtomGroup is created by running the `select_atoms` method on an Ensemble. It stores AtomGroup selections of the groups generated by running select atom on individual universes in a dictionary with the same key structure as the parent Ensemble class.

It returns a copy of the parent Ensemble object when  the `ensemble` method is run.

### EnsembleAnalysis

The EnsebmleAnalysis is a class inspired by the AnalysisBase in MDAnalyis which iterates over the systems in the ensemble and the frames in the systems. It sets up both iterations between universes and universe frames allowing for analysis to be run on both whole systems and the frames of those systems. This allows for users to easily run analyses on MDPOW simulations.

Example workflow
```python
    class DihedralAnalysis(mdpow.ensemble.EnsembleAnalysis):
        def __init__(self, DihedralEnsembleGroup):
            super(DihedralAnalysis, self).__init__(DihedralEnsembleGroup.ensemble())

            self._sel = DihedralEnsembleGroup

        def _prepare_ensemble(self):
            self.result_dict = {}
            for s in ['water', 'octanol']:
                self.result_dict[s] = {'Coulomb': {},
                                       'VDW': {}}
            for key in self._sel.group_keys():
                self.result_dict[key[0]][key[1]][key[2]] = None

        def _prepare_universe(self):
            self.angle_dict = {'angle': None,
                               'time': None}
            self.angles = []

        def _single_frame(self):
            angle = calc_dihedrals(self._sel[self._key].positions[0], self._sel[self._key].positions[1],
                                   self._sel[self._key].positions[2], self._sel[self._key].positions[3])
            self.angles.append(angle)

        def _conclude_universe(self):
            self.angle_dict['time'] = self.times
            self.angle_dict['angle'] = self.angles
            self.result_dict[self._key[0]][self._key[1]][self._key[2]] = self.angle_dict

        def _conclude_ensemble(self):
            self.results = pd.DataFrame(data=self.result_dict)
```

ALescoulie commented:
Since we're planning on dropping python 2 support #177 I did not go out of my way to make the python 2 cross compatible.

orbeckst commented:
It says that you changed 583 files. Is this true?

orbeckst commented:
Have you integrated the docs already, i.e., added an entry to [doc/sphinx/source/index.txt](https://github.com/Becksteinlab/MDPOW/blob/develop/doc/sphinx/source/index.txt) and an a txt file that autodocs the module? I also recommend you make an `analysis` subdirectory in the docs where you can start adding documentation for the analysis itself.

I am commenting here because the diff is currently to noisy to review.

ALescoulie commented:
@orbeckst I will defiantly start working on documentation, I understand that this is a large amount of new code, I'm also working on building more robust testing. Writing documentation will also coincide well with my report.

orbeckst commented:
Good that the docs builds but check https://mdpow--179.org.readthedocs.build/en/179/analysis.html 
- reST is not markdown  ;-) 
- code documentation is not included yet

see https://www.sphinx-doc.org/en/master/usage/quickstart.html and of course look at existing code and docs.

I also build the docs locally with
```bash
conda install sphinx sphinx_rtd_theme    # only needed once
python setup.py build_sphinx
open build/sphinx/html/index.html    # works on macOS
```

orbeckst commented:
>  I will *defiantly* start working on documentation

Oh-oh... ;-)

ALescoulie commented:
@orbeckst just got the documentation pushed. Tried to base it off of other documentation I had seen on the site.

ALescoulie commented:
Also the number of files changed if from the sloppy testing setup, I added a complete benzene simulation and have not gone through and set up more refined tests yet so there are uncompressed xtc files in the testing dir.

orbeckst commented:
Reduce the size of the test datasets
- use something with at least 1 hydrogen bond instead of benzene
- 3 Coul, 4 VDW windows
- trajectories only need to contain 4 frames
- try to keep size < 2 MB

orbeckst commented:
I merged current develop, so this is now definitely Python 3 only.

ALescoulie commented:
> I merged current develop, so this is now definitely Python 3 only.

Thanks, This makes the process of merging a bit simpler. I've spent today running new tests on Tubiwan.

orbeckst commented:
I merged develop into your branch, don't forget to `git pull` so that the history remains clean.

ALescoulie commented:
@orbeckst I have files for water, but for octanol I'm getting some failures on the NPT. I know we wanted to have two solvents for testing, so that right now is the biggest delay. Aside from that I have two other issues to fix.

- [ ] Add way to specify topology with kwarg
- [ ] Increase coverage especially for universe building

orbeckst commented:
What's the NPT failure?

ALescoulie commented:
I posted it in free energy methods in slack. It was due to pressure scaling greater than 1%

ALescoulie commented:
@orbeckst and @VOD555 I think this request is done. I still would like to get the octanol test files added, but at this point am having gromacs issues running it, so will add them later. I also added a kwarg for specifying a topology, and tested it. The one big issue still remaining is the lack of a test on the exception handling on the _load_dir_unv function in Ensemble, but I don't anticipate multiple topologies per lambda directory to be a issue for most use cases.

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,429 @@
+
+    Typical workflow for MDPOW directory::
+
+        ens = Ensemble(dirname='molecule')
```
It does that automatically during the __init__ if dirname is not none.

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,429 @@
+    Typical workflow for manual addition::
+
+        ens = Ensemble()
+        ens.add_system(topology='md.gro', trajectory='md.xvg')
```
fixed thanks for catching that one.

ALescoulie commented:
@orbeckst @VOD555 I think I addressed the changes both of you requested.

orbeckst commented on CHANGES:
```diff
@@ -21,7 +21,9 @@ Enhancements
   default None) and POWConfigParser.get() now logs when an option is used
   at level DEBUG (PR #187)
-
+* new Ensemble and EnsembleAtomGroup objects for loading set of system
+ simulations
+* new EnsembleAnalysis framework for collecting data from MDPOW simulations
```
indentation and include PR
```suggestion
* new Ensemble and EnsembleAtomGroup objects for loading set of system
  simulations (PR #179)
* new EnsembleAnalysis framework for collecting data from MDPOW simulations (PR #179)
```

ALescoulie replied:
done

orbeckst commented on mdpow/tests/test_ensemble.py:
```diff
@@ -0,0 +1,115 @@
+class TestEnsemble(object):
+    def setup(self):
+        self.tmpdir = td.TempDir()
+        self.old_path = os.getcwd()
+        self.resources = os.path.join(
+            self.old_path, 'mdpow', 'tests', 'testing_resources')
```
still todo

ALescoulie replied:
done

orbeckst commented on mdpow/tests/test_ensemble.py:
```diff
@@ -0,0 +1,115 @@
+        for k in Sol2_pos:
+            assert np.shape(Sol2_pos[k]) == (3, 3)
+        assert not Sol1 == Sol2
+        assert type(Sol2) == EnsembleAtomGroup
```
Please discuss.

ALescoulie replied:
done, wanted to ensure that it select_atoms() was correctly building an EnsemblAtomGroup

orbeckst commented on doc/sphinx/source/index.txt:
```diff
@@ -141,6 +141,8 @@ For current issues and open feature requests please look through the
    fep
    utilities
    forcefields
+   ensemble
+   analysis
```
ensemble should be inside analysis

ALescoulie replied:
@orbeckst think I got the docs updated to what you wanted.

orbeckst commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,34 @@
+systems to be loaded into `MDAnalysis <https://mdanalysis.org>`_
+Universes and be analyzed as a group.
+
+Ensemble Analysis Framework
+===========================
+
+EnsembleAnalysis
+----------------
+
+A framework for analyzing MDPOW simulations loaded into ensemble objects.
+
+.. autoclass:: mdpow.analysis.ensemble.EnsembleAnalysis
+    :members:
+
+    .. automethod:: _prepare_ensemble
+
+    .. automethod:: _prepare_universe
+
+    .. automethod:: _single_universe
+
+    .. automethod:: _single_frame
+
+    .. automethod:: _conclude_ensemble
+
+    .. automethod:: _conclude_universe
```
all of this should be in `analysis/ensemble.txt`

ALescoulie replied:
done

orbeckst commented on doc/sphinx/source/ensemble.txt:
```diff
@@ -0,0 +1,22 @@
+Ensemble Objects
```
move to `analysis/ensemble.txt` — these are the docs for the ensemble framework on which the analysis classes rely. Look at the MDA docs https://docs.mdanalysis.org/stable/documentation_pages/analysis_modules.html for analysis that list "building blocks" and then the actual analysis modules.

ALescoulie replied:
done

orbeckst commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,34 @@
+the objects in mdpow.ensemble allow for a molecule directory's
+systems to be loaded into `MDAnalysis <https://mdanalysis.org>`_
+Universes and be analyzed as a group.
+
```
Add a toc tree here, like
```reST
.. toctree::
   :maxdepth: 1
   :numbered:
   :hidden:

   analysis/ensemble
```
and then other txt files inside `analysis/` can be added later.

ALescoulie replied:
done

orbeckst commented on mdpow/tests/test_ensemble.py:
```diff
@@ -0,0 +1,111 @@
+RESOURCES = py.path.local(resource_filename(__name__, 'testing_resources'))
+MANIFEST = RESOURCES.join("manifest.yml")
+
+ensemble_keys = [('water', 'Coulomb', '0000'),
```
add a comment that we want to add tests for octanol later

ALescoulie replied:
done

orbeckst reviewed (changes requested):
Many of my previous comments are not visible on `mdpow/analysis/ensemble.py` but you can see them in the PR above on `mdpow/ensemble.py`. Please address those in addition to the new one. Thanks.

orbeckst commented:
@ALescoulie please add a comment to each of my comments when you fix it — either done or a reason why you don't want to or can't address it. This will make it easier for me when reviewing.  Don't *resolve* them, I'll do this when I re-review. Thanks.

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+=================
+ Analysis modules
+=================
+
```
added

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+.. Note:: The analysis module is a feature of 0.8.0 which only supports
+          python 3. It will still analyze existing simulations, but is not
+          included in existing python 2 MDPOW environment.
+
```
done

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+
+Ensemble
+________
+
```
added a sentence in the new ensemble.txt

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+
+EnsembleAtomGroup
+_________________
+
```
Added a sentence in ensemble.txt

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+
+EnsembleAnalysis
+----------------
+
```
Added a sentence to the analysis.txt

ALescoulie commented on mdpow/tests/testing_resources/manifest.yml:
```diff
@@ -40,3 +40,8 @@ analysis:
+    options:
+        - full_transfer
+    files:
+        - []
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+
+"""
+MDPOW.ensemble -- objects for analyzing MDPOW simulations
+========================================================
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+import MDAnalysis as mda
+from MDAnalysis.lib.log import ProgressBar
+from MDAnalysis.exceptions import FileFormatWarning, NoDataError, MissingDataWarning, SelectionError
+import numpy as np
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+EnsembleAnalysis:
+-----------------
+
+The EnsembleAnalysis is a class inspired by the AnalysisBase in MDAnalysis which iterates over the systems
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+EnsembleAtomGroup:
+-----------------
+
+The EnsembleAtomGroup is created by running the select_atoms method on an Ensemble. It stores AtomGroup
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+Ensemble:
+---------
+
+The Ensemble object is a collection of MDAnalyis Universe objects.
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+# 2021 Alia Lescoulie
+
+"""
+MDPOW.ensemble -- objects for analyzing MDPOW simulations
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+# MDPOW: ensemble.py
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+
+
+class NoDataWarning(Exception):
+    pass
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+
+    """
+
+    def __init__(self, dirname=None, solvents=['octanol', 'water'], **kwargs):
```
done, made it way simpler

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+
+        ens = Ensemble()
+        ens.add_system(topology='md.gro', trajectory='md.xvg')
+
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+        except KeyError:
+            logger.error(KeyError("Invalid system key"))
+
+    def __iter__(self):
```
no I'll add one

ALescoulie replied:
I decided to just remove the iter method since its not used, and not needed in EnsembleAnalysis, also makes more sense to iterate by using the list of keys to record which system in being iterated on.

ALescoulie commented:
> @ALescoulie please add a comment to each of my comments when you fix it — either "done" or a reason why you don't want to or can't address it. This will make it easier for me when reviewing.  Don't *resolve* them, I'll do this when I re-review. Thanks.

Sorry when I commit a suggestion it resolves the comments. I'm not intentionally hitting resolve.

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+        for file in cur_dir:
+            if file.endswith('.xtc'):
+                trj.append(file)
+            if (file.endswith('.tpr') or file.endswith('gro')) and self.top is None:
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+            else:
+                system = mda.Universe(top[0], [os.path.abspath(p) for p in trj])
+            return system
+        except FileFormatWarning or MissingDataWarning or NoDataError or ValueError:
```
done

ALescoulie replied:
I made it a tuple

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+        if len(top) == 0 or len(trj) == 0:
+            logger.warning('No MD files detected in %s', os.curdir)
+            raise NoDataWarning
+        trj.sort()
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+    def build_ensemble(self):
+        """Goes through given mdpow directory and attempts to build universe in the lambda directories.
+
+        Run if dirname is passed into __init__. First enters FEP directory, then goes through solvent
```
No they shouldn't want to run this manually, I should have included an underscore in it.

ALescoulie replied:
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+        and interaction directories to search lambda directories for system files."""
+        fep_dir = os.path.join(self.ensemble_dir, 'FEP')
+        for solvent in self.solvents:  # Ugly set of loops, may have to find way to clean up
+            for dirs in ["Coulomb", "VDW"]:  # Attribute folder names
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+                int_dir = os.path.join(fep_dir, solvent, dirs)
+                if os.path.exists(int_dir):
+                    with in_dir(int_dir, create=False):  # Entering attribute folders
+                        logger.info("Searching %s directory for systems" % os.curdir)
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+                                    try:
+                                        self.ensemble[(solvent, dirs, file)] = self._load_dir_unv(solvent)
+                                    except NoDataWarning:
+                                        continue
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+            sys_list.append(self[key])
+        return sys_list
+
+    def _load_dir_unv(self, solvent):
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+        except FileFormatWarning or MissingDataWarning or NoDataError or ValueError:
+            logger.error('Multiple trajectories detected in %s attempting'
+                           ' to find correct one for topology', os.curdir)
+            raise NoDataWarning
```
done

ALescoulie commented on mdpow/ensemble.py:
```diff
@@ -0,0 +1,428 @@
+        """Goes through given mdpow directory and attempts to build universe in the lambda directories.
+
+        Run if dirname is passed into __init__. First enters FEP directory, then goes through solvent
+        and interaction directories to search lambda directories for system files."""
```
done

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+          python 3. It will still analyze existing simulations, but is not
+          included in existing python 2 MDPOW environment.
+
+Ensemble Objects
```
done

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,50 @@
+.. Note:: The analysis module is a feature of 0.8.0 which only supports
+          python 3. It will still analyze existing simulations, but is not
+          included in existing python 2 MDPOW environment.
+
```
done

ALescoulie commented:
@orbeckst I think I addressed all the changes you suggested. Sorry about the resolving issues. I was not intentionally resolving it, rater GitHub resolves issues when files are changed (or rebased) or a suggestion is committed. 

@VOD555 if you have any other changes don't hesitate mention them.

orbeckst commented on mdpow/tests/test_ensemble.py:
```diff
@@ -0,0 +1,113 @@
+class TestEnsemble(object):
+    def setup(self):
+        self.tmpdir = td.TempDir()
+        self.m = pybol.Manifest(str(RESOURCES / 'manifest.yml'))
```
Does this work with a `py.path.local`?

orbeckst replied:
(i don't remember, cool if it does)

orbeckst commented on mdpow/tests/test_ensemble.py:
```diff
@@ -0,0 +1,113 @@
+    def test_build_exception(self):
+        ens = Ensemble()
+        with in_dir(os.path.join(self.tmpdir.name, 'FEP', 'test_solv'), create=False):
+            with pytest.raises(NoDataWarning):
```
NoDataError

ALescoulie replied:
It warns as that is raised when a directory does not have files.

orbeckst reviewed (changes requested):
Thanks for the updates. The following are still to be addressed:

- some of the warnings should probably exceptions
- given that Ensemble will be a framework, we need to test the exceptions (basically, no line of code comes for free... needs tests, especially if lots of other code will rely on it)
- some of the gro files should be replaced by bzip2-compressed gro files, leave one uncompressed
- some doc issues

There are some things that I should have seen/mentioned earlier, sorry I didn't. But it's getting there — the overall code structure is solid.

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+
+    def _conclude_universe(self):
+        """Run after each trajectory is finished"""
+        pass
```
done

ALescoulie replied:
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+                           os.curdir)
+        if len(top) == 0 or len(trj) == 0:
+            logger.warning('No MD files detected in %s', os.curdir)
+            raise NoDataWarning
```
It just means a lambda directory is empty, and exits the function scope to get the next lambda directory from _build_ensemble.

ALescoulie commented on doc/sphinx/source/analysis/ensemble.txt:
```diff
@@ -0,0 +1,32 @@
+Ensemble
+________
+
+The :class:`Ensemble` object is a collection of :class:`MDAnalyis.Universe` objects.
```
done

ALescoulie commented on doc/sphinx/source/analysis/ensemble.txt:
```diff
@@ -0,0 +1,32 @@
+________
+
+The :class:`Ensemble` object is a collection of :class:`MDAnalyis.Universe` objects.
+It is intended to store the set of systems generated by running mdpow-fep.
```
done

ALescoulie commented on doc/sphinx/source/analysis/ensemble.txt:
```diff
@@ -0,0 +1,32 @@
+The :class:`Ensemble` object works by storing the systems in a dictionary and extending the functionality
+of an :class:`MDAnalysis.Universe` to a collection of universes. It when given a directory finds the simulation
+files, reads then loads them into a dictionary. The object can be indexed the same as a dictionary,
+and has methods analogous the the Universe object. The main one being select_atoms which returns
```
done

ALescoulie commented on doc/sphinx/source/analysis/ensemble.txt:
```diff
@@ -0,0 +1,32 @@
+of an :class:`MDAnalysis.Universe` to a collection of universes. It when given a directory finds the simulation
+files, reads then loads them into a dictionary. The object can be indexed the same as a dictionary,
+and has methods analogous the the Universe object. The main one being select_atoms which returns
+a EnsembleAtomGroup. An :class:`Ensemble`in its current form can also be built by manually adding and popping
```
done

ALescoulie commented on doc/sphinx/source/analysis.txt:
```diff
@@ -0,0 +1,14 @@
+.. toctree::
+   :maxdepth: 1
+   :numbered:
+   :hidden:
```
done

ALescoulie commented on doc/sphinx/source/analysis/ensemble.txt:
```diff
@@ -0,0 +1,32 @@
+================
+Ensemble Objects
+================
```
done

ALescoulie commented on doc/sphinx/source/analysis/ensemble_analysis.txt:
```diff
@@ -0,0 +1,37 @@
+ Ensemble Analysis Framework
+============================
+
+.. versionadded:: 0.8.0
```
done

ALescoulie commented on doc/sphinx/source/index.txt:
```diff
@@ -141,7 +141,7 @@ For current issues and open feature requests please look through the
    utilities
    forcefields
-
+   analysis
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+    """Collection of related MDAnalysis Universe objects
+
+    Stores systems produced by running mdpow-fep organized
+    by solvent, interaction, and lambda.
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+
+        ens = Ensemble(dirname='molecule', topology_paths={'water': water_path,
+                                                           'octanol': octanol_path}
+
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            return
+
+        if not os.path.exists(dirname):
+            logger.error(f"Directory {dirname} does not exist")
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            item = self.ensemble[index]
+            return item
+        except KeyError:
+            logger.error(KeyError("Invalid system key"))
```
removed done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            if file.endswith('.tpr') and self.top is None:
+                # Ensures .tpr is used first if available
+                top = [file] + top
+            if (file.endswith('gro') or file.endswith('gro.b2z') or file.endswith('gro.gz')) and self.top is None:
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+
+        logs warning if more than one topology is in directory. If
+        more than one trajectory attempts to load both of them
+        in a universe if that fail will try to load each individually"""
```
done

orbeckst replied:
Still needs to be added to doc string.

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+                trj.append(file)
+            if file.endswith('.tpr') and self.top is None:
+                # Ensures .tpr is used first if available
+                top = [file] + top
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            if (file.endswith('gro') or file.endswith('gro.b2z') or file.endswith('gro.gz')) and self.top is None:
+                top.append(file)
+        if len(top) > 1:
+            logger.warning('More than one topology detected in %s will use first topology',
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            raise NoDataWarning
+
+    def get_keys(self):
+        return self.system_keys
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+        elif topology is None or trajectory is None:
+            err = f"{topology} does not exist"
+            logger.error(err)
+            raise ValueError(err)
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+        elif not os.path.exists(trajectory) and not os.path.exists(topology):
+            err = f"{trajectory} does not exist"
+            logger.error(err)
+            raise ValueError(err)
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+        self.system_keys.append(key)
+        self.num_systems += 1
+
+    def pop_system(self, key):
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+        except KeyError as err:
+            logger.error("%s with key %r" % (err, key))
+
+    def select_atoms(self, selection):
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            try:
+                ag = self[key].select_atoms(selection)
+                selections[key] = ag
+            except SelectionError as err:
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+        if isinstance(group_dict, dict):
+            self.keys = group_dict.keys()
+        else:
+            self.keys = None
```
no it will always be a dictionary

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            item = self.groups[index]
+            return item
+        except KeyError:
+            logger.error(KeyError("Invalid system key"))
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+                    return False
+            return True
+        else:
+            return False
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            return False
+
+    def __len__(self):
+        return len(self.group_keys())
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+    def __len__(self):
+        return len(self.group_keys())
+
+    def group_keys(self):
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+    def select_atoms(self, selection):
+        """Returns atom groups for Universe from objects
+
+        Presumes that solute in systems is the same"""
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+                selections[key] = ag
+            except SelectionError as err:
+                logger.warning("%s on system %r" % (err, key))
+                continue
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+    class in MDAnalysis <https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html>`_
+    and is a template for creating multiuniverse
+    multiframe analyses using the Ensemble object
+
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            Run on each universe in the ensemble during when
+            self.run in called.
+        """
+        pass
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+
+        Called on each frame for universes in the Ensemble.
+        """
+        pass
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+            def _conclude_ensemble(self):
+                self.results = pd.DataFrame(data=self.result_dict)
+
+        DihedralAnalysis.run(start=0 stop=10, step=1)
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+        """
+        logger.info("Setting up systems")
+        self._prepare_ensemble()
+        for self._key in self._ensemble.get_keys():
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,398 @@
+                self._setup_system(self._key, start=start, stop=stop, step=step)
+                self._prepare_universe()
+                self._single_universe()
+                for i, ts in enumerate(ProgressBar(self._trajectory[self.start:self.stop:self.step], verbose=True)):
```
done

ALescoulie commented:
@orbeckst I think I addressed all of your comments. I didn't think maintaining a website would be part of my work, but I got the docs updated to include everything. I also nearly maxed out the coverage, with just a few statements having partial coverage. In addition I added the custom progress bar.

ALescoulie commented:
@VOD555 Could you review when you get the chance. I think it's all done unless I made errors in the docs.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+                            if os.path.isdir(file):
+                                with in_dir(file, create=False):
+                                    try:
+                                        self._load_universe_from_dir(solvent)
+                                        self.add_system((solvent, dirs, file), topology=self.top[0],
+                                                        trajectory=self.trj)
```
Change to something along the lines of
```python
u = self._load_universe_from_dir(solvent=solvent)
if u is None:
   logger.warn(...)
   warnings.warn(..., category=MissingDataWarning)
key = (solvent, dirs, file)
self.add_system(key, u)
```

If there are exceptions to be caught then better use individual try/except blocks.

ALescoulie replied:
Done, now all the errors and exceptions are done in `_load_unviverse_from_dir()`

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+            self._keys.append(key)
+            self._num_systems += 1
+            return
+        if isinstance(trajectory, list):
+            # setting abs paths for trajectory
+            trajectory = [os.path.abspath(trj) for trj in trajectory]
```
move into `load_universe_from_dirs()`

ALescoulie replied:
done, I'll add a note in the docs as well so users know to do this too.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+            # setting abs paths for trajectory
+            trajectory = [os.path.abspath(trj) for trj in trajectory]
+        try:
+            self._ensemble[key] = mda.Universe(os.path.abspath(topology), trajectory)
```
Instead of the try/except block, this just becomes
```python
self._ensemble[key] = universe
```

ALescoulie replied:
done

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+            trajectory = [os.path.abspath(trj) for trj in trajectory]
+        try:
+            self._ensemble[key] = mda.Universe(os.path.abspath(topology), trajectory)
+        except (ValueError, FileFormatWarning, NoDataError, MissingDataWarning, OSError) as err:
```
... and the whole thing can go in the rewrite where this method just takes a universe as input

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+
+        self.trj = self._sort_trajectories(self.trj)
+        if len(self.top) > 1:
+            self.top = self._sort_topologies(self.top)
```
Now 
1. make the trajectory paths absolute
2. load the universe

If you have to protect the Universe creation with try/except, please also display the original exception so that one has a chance to debug things. In fact, it's probably best to just not do anything here and let any exceptions just happen. Start with 

```python
trajectory = [os.path.abspath(trj) for trj in util.asiterable(trajectory)]   # asiterable() _may_ be in util... or MDA?
return mda.Universe(topology, trajectory, **universe_kwargs)
```

Pass `universe_kwargs` through so that you have a fighting chance to modify universe creation.

ALescoulie replied:
done, added the path addition to `sort_topologies` and pass the kwargs

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+                break
+        return top
+
+    def _load_universe_from_dir(self, solvent):
```
Something like
```python
def _load_universe_from_dir(self, solvent=None, **universe_kwargs):
```
where `universe_kwargs` are any kwargs for MDAnalysis.Universe.

ALescoulie replied:
done

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+    def select_systems(self, keys=None, solvents=None, interactions=None,
+                       lambdas=None, lambda_range=None):
+        """Select specific subset of systems and returns them in an Ensemble.
+
```
add the parameter list here, then examples

ALescoulie replied:
done

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+
+            Ens = Ensemble(dirname='Mol')
+            w_v = Ens.select_systems(solvents=['water'], interactions=['VDW'],
+                                     lambda_range=[0, 1])
```
lambda_range is fragile (because of our simplistic naming convention) but convenient. Just a comment — let's keep it for MDPOW.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+        specifying the desired system attributes solvents, interactions and
+        lambdas. All arguments are stored in list form.
+
+        Specific key workflow example::
```
This looks good and should work for MDPOW.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+                                    new_key.append((s, i, k[2]))
+        for k in new_key:
+            logger.info('adding system %r to ensemble', k)
+            new_ens.add_system(k, universe=self[k])
```
nice, that's how I envisage add_system() being used

ALescoulie replied:
thanks, I'm glad Ensemble has been reasonable modular.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+
+    def __eq__(self, other):
+        if self.keys() == other.keys():
+            for k in self.keys():
+                if self[k] != other[k]:
+                    return False
+            return True
```
rewrite as 
```suggestion
            return all(self[k] == other[k] for k in self.keys())
```

ALescoulie replied:
done

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+                if self[k] != other[k]:
+                    return False
+            return True
+        else:
+            return False
```
can be shortened
```suggestion
       return False
```

ALescoulie replied:
done

orbeckst reviewed (changes requested):
Thank you for addressing my previous comments.

My main issue is with `_load_universe_from_dirs()` and `add_system()`. As you see in the comments, I'd like `_load_universe_from_dirs()` to be a pure function (method) without side effects that simply returns a Universe. Then `add_system()` can be simplified to _only_ add a Universe. This will make the code clearer and more robust.

It will also hopefully create less necessity for checking for NoDataError etc. In particular, get rid of the _exception_ `NoDataWarning`. If we need an exception for this case, use `NoDataError`. See the comments — apologies, the comments are a bit sprawling as I was learning how the code worked.

Compress almost all gro files as bz2, and leave one as gro and one as gz.

`select_systems()` looks good for MDPOW's purposes: it can select for what we care and does not do anything that's too fancy. I like how you can create new Ensembles and EnsembleAtomGroups. This looks like the right approach.

Regarding rewriting the history of this PR: Let's finish it here where we have all the comments. Once the PR is approved, you can make a new, clean one if you like. But I can also just squash this one, which will get rid of intermediate commits, too.

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+
+        if not os.path.exists(dirname):
+            logger.error(f"Directory {dirname} does not exist")
+            raise FileNotFoundError
```
The idea was to raise an exception if the directory structure given when initializing ensemble did not exist, but don't think it's actually needed. If the user gives incorrect settings it will just give an error anyway.

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+# MDPOW: _ensemble.py
+# 2021 Alia Lescoulie
+
+import os
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+            self._ensemble[key] = mda.Universe(os.path.abspath(topology), trajectory)
+        except (ValueError, FileFormatWarning, NoDataError, MissingDataWarning, OSError) as err:
+            logger.warning('Error loading systems at %r', key)
+            raise NoDataWarning
```
I set it up to log an exception and raise NoDataError. In the `_load_system_from_dir()` scope, the error, files and directory are logged.

ALescoulie commented on mdpow/analysis/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# Analysis Subpackage
+# Alia Lescoulie (ALescoulie) 2021
+
+class NoDataWarning(Warning):
+    """Warning indicating that data is missing"""
+    pass
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+        more than one trajectory attempts to load both of them
+        in a universe if that fail will try to load each individually"""
+        cur_dir = os.listdir(os.curdir)
+        self.trj = []
+        self.top = []
```
I made `_load_universe_from_dir()` a static method and return a Universe or None

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+                    logger.error('%r directory does not exist', int_dir)
+                    raise NoDataError
+
+    def add_system(self, key, topology=None, trajectory=None, universe=None):
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+        self.trj = []
+        self.top = []
+
+        if not self.top_dict is None:
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+        self.top = []
+
+        if not self.top_dict is None:
+            # if top is specified in kwargs, saved to list
```
done

ALescoulie commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,499 @@
+
+        if not self.top_dict is None:
+            # if top is specified in kwargs, saved to list
+            self.top = [self.top_dict[solvent]]
```
I restructured that to be in the `_build_ensemble` scope. `self.top_dict` is set to whatever the `topology_paths` argument is which defaults to None if it's unassigned.

ALescoulie commented:
@orbeckst I finished addressing all your comments and got it near complete coverage on ensemble with the only thing being some partials.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,525 @@
+                int_dir = os.path.join(fep_dir, solvent, dirs)
+                with in_dir(int_dir, create=False):  # Entering attribute folders
+                    logger.info("Searching %s directory for systems", os.curdir)
+                    files: list = os.listdir(os.curdir)
```
Why the type annotation?

ALescoulie replied:
not really important, can be removed.

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,525 @@
+                    for file in sorted(files):  # Traversing lambda directories
+                        if os.path.isdir(file):
+                            with in_dir(file, create=False):
+                                u = self._load_universe_from_dir(solv_dir=solv_top_path,
+                                                                 **self.unv_kwargs)
+                                if u is None:
+                                    logger.warning(f'No system loaded in {file}')
+                                else:
+                                    self.add_system((solvent, dirs, file), u)
```
I like it!

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,525 @@
+    def add_system(self, key, universe: mda.Universe):
+        """Adds system from universe object for trajectory and topology files
+
+        Takes specified key and either existing mda.Universe object or
+        trajectory and topology path. Ensure that paths are set to absolute
```
update doc string

ALescoulie replied:
done

orbeckst commented on mdpow/analysis/ensemble.py:
```diff
@@ -0,0 +1,525 @@
+        self.frames = np.zeros(self.n_frames, dtype=int)
+        self.times = np.zeros(self.n_frames)
+
+    def _single_universe(self):
+        """Calculations on a single Universe object.
+
+            Run on each universe in the _ensemble during when
+            self.run in called.
+        """
+        pass  # pragma: no cover
```
how is `_single_universe` different from `_prepare_universe`? Do you need both?

ALescoulie replied:
`_prepare_universe` is setting up frame iteration, single universe is if the user wants to do some kind of method that runs on a full universe.

orbeckst reviewed (changes requested):
Very good changes, code is much cleaner and excellent test coverage improvement.

* A few minor comments (see inline), mostly docs.
* Compress almost all of the gro files with bz2, one with gz, and leave one uncompressed.

orbeckst commented:
@VOD555 can you please check if @ALescoulie satisfied your review requirements. If you're happy please approve, otherwise please ask for any other necessary changes.

ALescoulie commented:
@orbeckst I got the files compressed and uploaded.

VOD555 reviewed (approved):
I'm ok with the updates.

orbeckst commented:
I am working on the squash merge... will be merged in <5 mins.

orbeckst commented:
Hooray, it's in, congratulations @ALescoulie 🥳  !!!
